Feat/152/suspend user log

bradtaniguchi opened the pull request:
**Description:**
- Refactor schema/model setup with mongoose to reflect the new "userlog" feature, while making some changes to support pre-existing suspend logic

Related to #152 

**note** this is a draft since I want to get some feedback on how the schema is currently setup. I'm also not sure about how the `userId` isn't the same as the mongodb `_id` property, which could just be made to be the discord `snowflake` identifier instead. (its a rather long unique number)

Commit:
Feature/suspend logging (#153)

* Reconfigure suspend logging

* Modify user command

* Reformat log output

* Refactor display message for log

* Refactor

bradtaniguchi committed:
Merge branch 'master' into feat/152/suspend-user-log

bradtaniguchi committed:
Add new "refactored" version of db files

- created seperate schemas for the different models, along with a few
helper files to help keep track of stuff.

bradtaniguchi committed:
Add user-log model/schema

naomi-lgbt commented on src/db/moderator.ts:
```diff
@@ -0,0 +1,49 @@
+  /**
+   * The discord server nickname of the mod
+   */
+  nickname: string;
```
To clarify, the nickname can be `null`. You and I both do not have nicknames set, so this field is empty. To typecast as a string, I used the `member.nickname` field and a fallback of `""`.

bradtaniguchi replied:
Ah I didn't know that good call! I'll update the comments and schema to set this as an empty string as default.

naomi-lgbt commented on src/db/user.ts:
```diff
@@ -0,0 +1,39 @@
+   * users incase they change their discord user name.
+   */
+  identities: Array<UserIdentity>;
+}
```
So this will be a separate model from the user's suspension history?

bradtaniguchi replied:
So the idea I had was the `User` would be used to identify user's from Discord and possibly FCC itself. 

The suspension history/logging history/whatever information will be saved in other schemas by itself. This should make the database more flexible if we end up adding more stuff. Everything will have to have the same reference to the original user.

**I wasn't sure about the `userId` being different than the mongodb generated _id**, and was thinking if we should maybe make them 1 in the same?

naomi-lgbt replied:
Discord ID values are theoretically unique, so that shouldn't be a problem. The only concern I would have there is future-proofing (if Discord runs out of numbers or switches how they identify a user, we'd have to change the `_id` value and I'm not sure mongo likes that).

naomi-lgbt commented on src/db/moderator.ts:
```diff
@@ -0,0 +1,49 @@
+  userId: string | Snowflake;
+  /**
+   * The discord server nickname of the mod
+   */
```
On this one, I believe the `user` value is a Snowflake, but storing the `user.id` value is a string. We should type it as only one or the other to enforce consistency in the codebase, right?

bradtaniguchi replied:
I might of left this `Snowflake` in here on accident.

naomi-lgbt commented on src/db/user-suspension.ts:
```diff
@@ -0,0 +1,53 @@
+  user: string | ObjectId | User;
+  /**
+   * The reason the user was suspended.
+   */
```
Again, I think we need to determine what value will fill this field so we can be more strict with the typing.

naomi-lgbt reviewed (commented):
I ran these changes locally and because nothing has changed in the workflow yet I'm having trouble visualising how everything will come together. 😅 

I added some comments on a few things, but overall it looks good to me - I'm curious to see what will get saved when I suspend a user, and what I see when I call the `user` command.

bradtaniguchi committed:
Migrated userId to _id for snowflake

bradtaniguchi committed:
Add create-user-suspension logic

bradtaniguchi committed:
Update user-suspension logic

bradtaniguchi committed:
Removed nickname property from moderator

- this property isn't available from the `message.author` property from
  discord. We won't save it.

bradtaniguchi committed:
Add create-user-supension utility function

- this will create/update properties within the database.

bradtaniguchi committed:
Add create-user-log function

- moved logic from create-user-suspension to be re-usable.

bradtaniguchi committed:
Update un-used imports

- fixes issue reported on LGTM

bradtaniguchi committed:
Merge branch 'master' into feat/152/suspend-user-log

bradtaniguchi committed:
Fix lint error

bradtaniguchi committed:
Add get-user-entry function
